### ISSUE TITLE
Fan-In Pipeline Triggered Multiple Times When Upstream Pipeline Started Manually

### ISSUE THREAD

**Issue by noelob** [gocd/gocd#10537]
##### Issue Type

- Bug Report

##### Summary

Fan-in pipelines are run once when an upstream pipelines it triggered by a Git commit, but if the upstream pipeline is triggered manually, the fan-in pipeline is run multiple times (once for each of the fanned out pipelines). 

###### Basic environment details

* Go Version: `22.1.0`
* JAVA Version: `13.0.2`
* OS: `Mac OS X 10.16`

##### Steps to Reproduce

1. Download and run GoCD as per the [Test Drive GoCD page](https://www.gocd.org/test-drive-gocd.html). 
2. In the `Example-Pipelines-Fan-Out-Fan-In` group, run the `Build-Test` pipelines manually

##### Expected Results
Expected the `Deploy-Staging` pipeline to be executed once

##### Actual Results
The `Deploy-Staging` pipeline was executed twice, once for each of the fanned-out stages (`Acceptance-Tests` and `Performance-Tests`)

<img width="1091" alt="Screenshot 2022-06-20 at 15 25 14" src="https://user-images.githubusercontent.com/643338/174684380-e7780a72-9da1-4947-93bc-a2f4b547795d.png">

Note that when the same pipelines are triggered by a Git revision (as when the GoCD test drive starts up), `Deploy-Staging` is executed only once:
<img width="1089" alt="Screenshot 2022-06-20 at 15 27 14" src="https://user-images.githubusercontent.com/643338/174684502-26cb965f-2116-4c76-8292-b5fe57eb41ee.png">

##### Any other info

This looks related to https://github.com/gocd/gocd/issues/2225. There are comments on that issue that explain how the fan-in works based on the source material, and I think I understand those.

Why does manually triggering a pipeline result in different behaviour than when it's triggered via Git?
Is there a way to workaround this? If my fan-in stage is `deploy`, or `publish-to-artifactory`, those are not activities that I want to have run repeatedly.

**Comment by chadlwilson** [comment]
This is working as originally designed, even if it's perhaps not the behaviour you would prefer. The material-based fan-in ensures that where the inputs to a pipeline share upstream materials the upstream pipeline has successfully been run with the same material revisions.

When you re-run a pipeline stage, the upstream materials they were run with don't change, so for every completion of acceptance-tests and performance tests the _deploy-staging_ pipeline can trigger because the material version (from `test-repo` in example above) has not changed, and from _deploy-staging_s perspective already has compatible runs of both _acceptance-tests_ and _performance-tests_ that meet the fan-in requirements.

This logic is explained in https://github.com/gocd/gocd/issues/2225#issuecomment-284684317 and it seems your question is essentially a duplicate of that - this "root material driven" logic is why it becomess "triggered multiple times" from your perspective.

In general I would perhaps suggest
- artifact uploads/publishes should either be idempotent or done with the pipeline label version
- consider modelling of declarative/idempotent deploys or choosing tools that support that

This way, re-runs where the actual built artifact does not change lead to quick-running no-op deployments.

**Comment by chadlwilson** [comment]
Not in GoCD, to my knowledge. Just using that to distinguish between how some other tools attempt fan-in where I think they do it based on expected instances/triggers of the upstream pipelines (which has other bad effects sometimes).
